torch/nn/utils/parametrizations.py

```diff
@@ -1,10 +1,273 @@
+    ``"householder"``, but they are slower to compute for very thin or very wide matrices.
+
+    If ``extra_memory=True`` (default), the parametrization implements the "Dynamic Trivialization Framework",
+    where an extra matrix :math:`B \in \mathbb{K}^{n \times n}` is stored under
+    ``module.parametrizations.weight[0].base``. This helps the
+    convergence of the parametrized layer at the expense of some extra memory use.
+    See `Trivializations for Gradient-Based Optimization on Manifolds`_ .
+
+    .. note::
+        This function is implemented using the parametrization functionality
+        in :func:`torch.nn.utils.parametrize.register_parametrization`.
+
+    .. note ::
+        If the original tensor is not parametrized and ``extra_memory=True`` (default), the initial value
+        of ``Q`` is that of the original tensor if it is orthogonal (or unitary in the complex case)
+        and it is orthogonalized via the QR decomposition otherwise (see :func:`torch.linalg.qr`).
+        Same happens when ``parametrization="householder"`` even when ``extra_memory=False``.
+        Otherwise, the initial value is the result of the composition of all the registered
+        parametrizations applied to the original tensor.
+
+    .. _`Trivializations for Gradient-Based Optimization on Manifolds`: https://arxiv.org/abs/1909.09501
+
+    Args:
+        module (nn.Module): module on which to register the parametrization
+        name (str, optional): name of the tensor to make orthogonal. Default: ``"weight"``.
+        parametrization (str, optional): One of the following: ``["matrix_exp", "cayley", "householder"]``.
+            Default: ``"matrix_exp"`` if the matrix is square, ``"householder"`` otherwise.
+        extra_memory (bool, optional): whether to use the dynamic trivialization framework.
+            Default: ``True``.
+
+    Returns:
+        The original module with an orthogonal parametrization registered to the specified
+        weight
+
+    Example::
+
+        >>> orth_linear = orthogonal(nn.Linear(20, 40))
+        >>> orth_linear
+        ParametrizedLinear(
+        in_features=20, out_features=40, bias=True
+        (parametrizations): ModuleDict(
+            (weight): ParametrizationList(
+            (0): _Orthogonal()
+            )
+        )
+        )
+        >>> Q = orth_linear.weight
+        >>> torch.dist(Q.T @ Q, torch.eye(20))
+        tensor(4.9332e-07)
+    """
+    if not hasattr(module, name):
+        raise ValueError(
+            "Module '{}' has no attribute with name '{}'".format(module, name)
+        )
+    # getattr should get the correct parametrized weight if there
+    # is already an parametrization registered
+    weight = getattr(module, name)
+    if parametrization is None:
+        parametrization = "householder" if weight.size(-2) != weight.size(-1) else "matrix_exp"
+
+    if parametrization not in ["matrix_exp", "cayley", "householder"]:
```

Comment:
nit: maybe create a small union to contain these? That would make checking here easier and allow typing to catch any error.